Firestore/core/test/unit/model/values_test.cc

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/model/values.h"
+#include "Firestore/core/src/model/database_id.h"
+#include "Firestore/core/src/model/field_value.h"
+#include "Firestore/core/src/remote/serializer.h"
+#include "Firestore/core/src/util/comparison.h"
+#include "Firestore/core/test/unit/testutil/equals_tester.h"
+#include "Firestore/core/test/unit/testutil/testutil.h"
+#include "Firestore/core/test/unit/testutil/time_testing.h"
+#include "absl/base/casts.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+using testutil::Array;
+using testutil::BlobValue;
+using testutil::DbId;
+using testutil::Key;
+using testutil::Map;
+using testutil::time_point;
+using testutil::Value;
+using util::ComparisonResult;
+
+namespace {
+
+double ToDouble(uint64_t value) {
+  return absl::bit_cast<double>(value);
+}
+
+const uint64_t kNanBits = 0x7fff000000000000ULL;
+
+static time_point kDate1 = testutil::MakeTimePoint(2016, 5, 20, 10, 20, 0);
+static Timestamp kTimestamp1{1463739600, 0};
+
+static time_point kDate2 = testutil::MakeTimePoint(2016, 10, 21, 15, 32, 0);
+static Timestamp kTimestamp2{1477063920, 0};
+
+class ValuesTest : public ::testing::Test {
+ public:
+  ValuesTest() : serializer(DbId()) {
+  }
+
+  template <typename T>
+  google_firestore_v1_Value Wrap(T input) {
+    model::FieldValue fv = Value(input);
+    return serializer.EncodeFieldValue(fv);
+  }
+
+  template <typename... Args>
+  google_firestore_v1_Value WrapObject(Args... key_value_pairs) {
+    FieldValue fv = testutil::WrapObject((key_value_pairs)...);
```

Comment:
Nits:
* the extra parentheses around `key_value_pairs` are unnecessary;
* it's not very important because it's a test file, but ideally you would make sure to forward the arguments to `WrapObject` to avoid extra copies (though `WrapObject`'s implementation would then proceed to create those copies regardless -- ideally it should be fixed, but that's out of scope for this PR):
```cpp
  template <typename... Args>
  // Make `Args` a forwarding reference
  google_firestore_v1_Value WrapObject(Args&&... key_value_pairs) {
    FieldValue fv = testutil::WrapObject(std::forward<Args>(key_value_pairs)...);
```
Feel free to disregard the point re. forwarding.